src/common/linear_assignment_cache.h

```diff
@@ -0,0 +1,148 @@
+		      assignment_() {}
+
+		static bool partsMatch(const detail::Slice::PartProxy &part1,
+		                       const detail::Slice::PartProxy &part2) {
+			return part1.size() == part2.size() &&
+			       std::equal(part1.begin(), part1.end(), part2.begin());
+		}
+
+		template <std::size_t N>
+		bool checkMatchAndGetResult(Goal::Slice &target_slice,
+		                            Goal::Slice &source_slice,
+		                            std::array<int, N> &assignment) {
+			if (target_slice.getType() != target_slice_.getType() ||
+			    source_slice.getType() != source_slice_.getType()) {
+				return false;
+			}
+			assert(target_slice.size() == target_slice_.size());
+			assert(source_slice.size() == source_slice_.size());
+
+			for (int i = 0; i < target_slice.size(); ++i) {
+				if (!partsMatch(target_slice[i], target_slice_[i])) {
+					return false;
+				}
+			}
+
+			for (int i = 0; i < source_slice.size(); ++i) {
+				if (!partsMatch(source_slice[i], source_slice_[i])) {
+					return false;
+				}
+			}
+
+			// match is good
+			std::copy(assignment_.begin(), assignment_.end(),
+			          assignment.begin());
+
+			return true;
+		}
+
+		template <std::size_t N>
+		void store(Goal::Slice &target_slice, Goal::Slice &source_slice,
+		           std::array<int, N> &assignment) {
+			target_slice_ = target_slice;
+			source_slice_ = source_slice;
+			assignment_.assign(assignment.begin(), assignment.end());
+		}
+
+	private:
+		Goal::Slice target_slice_, source_slice_;
+		std::vector<int> assignment_;
+	};
+
+	using HashType = int64_t;
+	using EntriesContainer = std::unordered_map<HashType, OptimizerInputOutput>;
+
+	LinearAssignmentCache() : cachedData_() {}
+
+	HashType getLabelValue(MediaLabel label) const {
+		return static_cast<MediaLabelManager::HandleValue>(label);
+	}
+
+	HashType hashPart(const detail::Slice::PartProxy &part) const {
+		HashType hash = 1;
+		for (const auto &label : part) {
+			HashType label_hash =
+			    (getLabelValue(label.first) * Goal::kMaxExpectedCopies +
+			     label.second) %
+			    kBasePart_;
+			hash = (hash * kBasePart_ + label_hash) % kModPart_;
+		}
+		return hash;
+	}
+
+	HashType hashSlice(HashType startingHash, Goal::Slice &slice) const {
+		HashType hash = startingHash;
+		for (const auto &part : slice) {
+			HashType part_value = hashPart(part);
+			hash = (hash * kBaseSlice_ + part_value) % kModSlice_;
+		}
+		return hash;
+	}
+
+	template <std::size_t N>
+	bool checkMatchAndGetResult(Goal::Slice &target_slice,
+	                            Goal::Slice &source_slice,
+	                            std::array<int, N> &assignemt) {
+		HashType hash = hashSlice(kInitialStartingHash, target_slice);
+		hash = hashSlice(hash, source_slice);
+		if (cachedData_[hash].checkMatchAndGetResult<N>(
+		        target_slice, source_slice, assignemt)) {
+			return true;
+		}
+		return false;
+	}
+
+	template <std::size_t N>
+	void store(Goal::Slice &target_slice, Goal::Slice &source_slice,
```

Comment:
Typo in the signature of `void store(...)`: `assignemt` instead of `assignment`.